File: 2019/python/day10.py
#!python3

from collections import Counter, defaultdict
from itertools import combinations
from math import atan2, pi

from helpers import BASE
from maths import gcd

TEST01 = '''.#..#
.....
#####
....#
...##'''

TEST02 = '''......#.#.
#..#.#....
..#######.
.#.#.###..
.#..#.....
..#....#.#
#..#....#.
.##.#..###
##...#..#.
.#....####'''

TEST03 = '''#.#...#.#.
.###....#.
.#....#...
##.#.#.#.#
....#.#.#.
.##..###.#
..#...##..
..##....##
......#...
.####.###.'''

TEST04 = '''.#..#..###
####.###.#
....###.#.
..###.##.#
##.##.#.#.
....###..#
..#.#..#.#
#..#.#.###
.##...##.#
.....#.#..'''

TEST05 = '''.#..##.###...#######
##.############..##.
.#.######.########.#
.###.#######.####.#.
#####.##.#.##.###.##
..#####..#.#########
####################
#.####....###.#.#.##
##.#################
#####.##.###..####..
..######..##.#######
####.##.####...##..#
.#####..#.######.###
##...#.##########...
#.##########.#######
.####.#.###.###.#.##
....##.##.###..#####
.#.#.###########.###
#.#.#.#####.####.###
###.##.####.##.#..##'''


def parse(string):
    asteroids = set([])
    for y, line in enumerate(string.strip().split('\n')):
        for x, ch in enumerate(line):
            if ch == '#':
                asteroids.add((x, y))
    return asteroids


def make_vectors(asteroids):
    vectors = defaultdict(list)
    for a1, a2 in combinations(asteroids, 2):
        x1, y1 = a1
        x2, y2 = a2
        vectors[a1].append((x2 - x1, -(y2 - y1)))
        vectors[a2].append((x1 - x2, -(y1 - y2)))
    return vectors


def make_angles(vectors, best):
    angles = defaultdict(list)
    for x, y in vectors:
        a = atan2(y, x)
        if a < 0:
            # Switch the angles below the x-axis to rotate anti-clockwise.
            a += 2 * pi
        angles[a].append((x, y))
    # Switch to clockwise rotation.
    angles = {2 * pi - k if k > 0 else k: v for k, v in angles.items()}
    # Make zero angle slightly greater than zero.
    angles = {(k + 0.000001) if k == 0 else k: v for k, v in angles.items()}
    # Rotate to make the positive y-axis the zero angle.
    angles = {(k + pi / 2) % (2 * pi): v for k, v in angles.items()}
    for a, ast in angles.items():
        ast.sort(key=lambda x: abs(x[0]) + abs(x[1]))
    return angles



def test1():
    asteroids = parse(TEST01)
    assert part1(asteroids) == (8, (3, 4))


def test2():
    asteroids = parse(TEST02)
    assert part1(asteroids)[0] == 33


def test3():
    asteroids = parse(TEST03)
    assert part1(asteroids)[0] == 35


def test4():
    asteroids = parse(TEST04)
    assert part1(asteroids)[0] == 41


def test5():
    asteroids = parse(TEST05)
    assert part1(asteroids) == (210, (11, 13))


def test6():
    asteroids = parse(TEST05)
    _, best = part1(asteroids)
    assert part2(asteroids, best) == 802


def part1(asteroids):
    vs = make_vectors(asteroids)
    counter = Counter()
    for asteroid, vectors in vs.items():
        unique_vectors = set()
        for v in vectors:
            x = gcd(*v)
            if x != 1:
                v = (v[0] // x, v[1] // x)
            unique_vectors.add(v)
        counter[asteroid] = len(unique_vectors)
    m = counter.most_common(1)[0]
    return m[1], m[0]


def part2(asteroids, best):
    vs = make_vectors(asteroids)
    best_vectors = vs[best]
    angles = make_angles(best_vectors, best)
    count = 0
    sorted_angles = sorted(angles.keys())
    while True:
        for angle in sorted_angles:
            ast = angles[angle]
            try:
                a = ast.pop(0)
                count += 1
                if count == 200:
                    return 100 * (best[0] + a[0]) + best[1] - a[1]
            except IndexError:
                pass


def main():
    test1()
    test2()
    test3()
    test4()
    test5()

    asteroids = parse(open(BASE + 'day10.txt', 'r').read().strip())

    p1, best = part1(asteroids)
    print(f'Part 1: {p1}')

    test6()

    p2 = part2(asteroids, best)
    print(f'Part 2: {p2}')


if __name__ == '__main__':
    main()

File: 2019/swift/AdventOfCode2019a/AdventOfCode2019a/Day10.swift
//
//  Day10.swift
//  AdventOfCode2019a
//
//  Created by gary on 15/01/2023.
//

import Foundation

private typealias Map = [Point: [Point: (Point, Double)]]
private typealias Part2 = [Point: (Point, Double)]


func day10() {
//    let s = test5.trimmingCharacters(in: .whitespacesAndNewlines)
//        .split(separator: "\n")
//        .map({ String($0) })
    let s = get(file: "10")
        .split(separator: "\n")
        .map({ String($0) })

    let points = makePoints(map: s)
    let map = makeMap(points: points)

    let (p1, bestPoint) = part1(map: map)
    print(p1)

    let data = map[bestPoint]!
    let p2 = part2(point: bestPoint, data: data)
    print(p2)
}


private func part1(map: Map) -> (Int, Point) {
    var bestCount = 0
    var bestPoint = Point.origin
    for (p, values) in map {
        var s: Set<Point> = []
        for (point, _) in values.values {
            s.insert(point)
        }
        if s.count > bestCount {
            bestCount = s.count
            bestPoint = p
        }
    }
    return (bestCount, bestPoint)
}


private func part2(point: Point, data: Part2) -> Int {
    var result: [Double: [(Double, Point)]] = [:]
    for (k, v) in data {
        let (vector, d) = v
        let a = getAngle(point: vector)
        result[a, default: []].append((d, k))
    }
    let angles = result.keys.sorted()
    var counter = 0
    var rotation = 1
    while true {
        for angle in angles {
            let items = result[angle]!
            if items.count >= rotation {
                counter += 1
            }
            if counter == 200 {
                let p = items[rotation - 1].1
                return 100 * p.x + p.y
            }
        }
        rotation += 1
    }
}


private func makePoints(map: [String]) -> [Point] {
    var points: [Point] = []
    for (j, row) in map.enumerated() {
        for (i, ch) in row.enumerated() {
            if ch == "#" {
                let p = Point(x: i, y: j)
                points.append(p)
            }
        }
    }
    return points
}


private func makeMap(points: [Point]) -> [Point: [Point: (Point, Double)]] {
    var map: [Point: [Point: (Point, Double)]] = [:]
    for (i, p1) in points.enumerated() {
        for p2 in points[(i + 1)...] {
            let d = hypot(p1: p1, p2: p2)
            var v = p2 - p1
            let g = gcd(a: v.x, b: v.y)
            v = v / g
            map[p1, default: [:]][p2] = (v, d)
            map[p2, default: [:]][p1] = (-1 * v, d)
        }
    }
    return map
}


private func getAngle(point: Point) -> Double {
    let a = atan2(Double(point.y), Double(point.x))
    var b = a < 0 ? 2 * Double.pi + a : a
    b += Double.pi / 2
    if b >= 2 * Double.pi {
        return b - 2 * Double.pi
    } else {
        return b
    }
}


private let test1 = """
.#..#
.....
#####
....#
...##
"""

private let test2 = """
......#.#.
#..#.#....
..#######.
.#.#.###..
.#..#.....
..#....#.#
#..#....#.
.##.#..###
##...#..#.
.#....####
"""

private let test3 = """
#.#...#.#.
.###....#.
.#....#...
##.#.#.#.#
....#.#.#.
.##..###.#
..#...##..
..##....##
......#...
.####.###.
"""

private let test4 = """
.#..#..###
####.###.#
....###.#.
..###.##.#
##.##.#.#.
....###..#
..#.#..#.#
#..#.#.###
.##...##.#
.....#.#..
"""

private let test5 = """
.#..##.###...#######
##.############..##.
.#.######.########.#
.###.#######.####.#.
#####.##.#.##.###.##
..#####..#.#########
####################
#.####....###.#.#.##
##.#################
#####.##.###..####..
..######..##.#######
####.##.####...##..#
.#####..#.######.###
##...#.##########...
#.##########.#######
.####.#.###.###.#.##
....##.##.###..#####
.#.#.###########.###
#.#.#.#####.####.###
###.##.####.##.#..##
"""
